emcc.py

```diff
@@ -1481,6 +1484,31 @@ def check(input_file):
     if shared.Settings.MIN_CHROME_VERSION <= 37:
       shared.Settings.WORKAROUND_OLD_WEBGL_UNIFORM_UPLOAD_IGNORED_OFFSET_BUG = 1
 
+    # Environment setting based on user input
+    environments = shared.Settings.ENVIRONMENT.split(',')
+    if any([x for x in environments if x not in VALID_ENVIRONMENTS]):
+      exit_with_error('Invalid environment specified in "ENVIRONMENT": ' + shared.Settings.ENVIRONMENT + '. Should be one of: ' + ','.join(VALID_ENVIRONMENTS))
+
+    shared.Settings.ENVIRONMENT_MAY_BE_WEB = not shared.Settings.ENVIRONMENT or 'web' in environments
+    shared.Settings.ENVIRONMENT_MAY_BE_WEBVIEW = not shared.Settings.ENVIRONMENT or 'webview' in environments
+    shared.Settings.ENVIRONMENT_MAY_BE_NODE = not shared.Settings.ENVIRONMENT or 'node' in environments
+    shared.Settings.ENVIRONMENT_MAY_BE_SHELL = not shared.Settings.ENVIRONMENT or 'shell' in environments
+
+    # The worker case also includes Node.js workers when pthreads are
+    # enabled and Node.js is one of the supported environments for the build to
+    # run on. Node.js workers are detected as a combination of
+    # ENVIRONMENT_IS_WORKER and ENVIRONMENT_IS_NODE.
+    shared.Settings.ENVIRONMENT_MAY_BE_WORKER = \
+        not shared.Settings.ENVIRONMENT or \
+        'worker' in environments or \
+        (shared.Settings.ENVIRONMENT_MAY_BE_NODE and shared.Settings.USE_PTHREADS)
+
+    if not shared.Settings.ENVIRONMENT_MAY_BE_WORKER and shared.Settings.PROXY_TO_WORKER:
+      exit_with_error('If you specify --proxy-to-worker and specify a "-s ENVIRONMENT=" directive, it must include "worker" as a target! (Try e.g. -s ENVIRONMENT=web,worker)')
+
+    if not shared.Settings.ENVIRONMENT_MAY_BE_WORKER and shared.Settings.USE_PTHREADS:
+      exit_with_error('When building with multithreading enabled and a "-s ENVIRONMENT=" directive is specified, it must include "worker" as a target! (Try e.g. -s ENVIRONMENT=web,worker)')
```

Comment:
Can you make this a new function rather than adding more to run()?

Reply:
Good idea, done.